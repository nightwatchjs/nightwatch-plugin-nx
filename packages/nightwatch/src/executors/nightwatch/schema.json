{
  "title": "nightwatch Target",
  "description": "Run nightwatch for e2e, integration and component testing.",
  "type": "object",
  "outputCapture": "pipe",
  "cli": "nx",
  "presets": [
    {
      "name": "Starting Dev Server",
      "keys": [
        "nightwatchConfig",
        "devServerTarget"
      ]
    },
    {
      "name": "Custom Base Url",
      "keys": [
        "nightwatchConfig",
        "baseUrl"
      ]
    }
  ],
  "properties": {
    "nightwatchConfig": {
      "type": "string",
      "description": "The path of the nightwatch configuration json file.",
      "x-completion-type": "file",
      "x-completion-glob": "nightwatch.conf?(*)@(.js|.ts|.json)"
    },
    "tsConfig": {
      "type": "string",
      "description": "The path of the nightwatch tsconfig configuration json file.",
      "x-completion-type": "file",
      "x-completion-glob": "tsconfig.*.json"
    },
    "devServerTarget": {
      "type": "string",
      "description": "Dev server target to run tests against."
    },
    "parallel": {
      "type": "boolean",
      "description": "Whether or not nightwatch should run its tests in parallel.",
      "default": false
    },
    "baseUrl": {
      "type": "string",
      "description": "The address (with the port) which your application is running on."
    },
    "env": {
      "type": "object",
      "description": "A key-value Pair of environment variables to pass to nightwatch runner."
    },
    "spec": {
      "type": "string",
      "description": "A comma delimited glob string that is provided to the nightwatch runner to specify which spec files to run. i.e. `**examples/**,**actions.spec**`."
    },
    "reporter": {
      "type": "string",
      "description": "The reporter used during nightwatch run."
    },
    "reporterOptions": {
      "type": "string",
      "description": "The reporter options used. Supported options depend on the reporter."
    },
    "tag": {
      "type": "string",
      "description": "A comma delimited list to identify a run with.",
      "aliases": [
        "t"
      ]
    }
  },
  "additionalProperties": true,
  "required": [
    "nightwatchConfig"
  ],
  "examplesFile": "../../../docs/nightwatch-examples.md"
}
