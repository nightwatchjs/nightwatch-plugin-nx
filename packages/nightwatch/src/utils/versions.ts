export const nxVersion = require('../../package.json').version;
export const typesNodeVersion = '16.11.7';
export const nightwatchVersion = '^2.5.0';
export const typesNightwatchVersion = '2.3.12';
export const geckodriver = "3.2.0";
export const chromedriver = "107.0.3";
export const webpackHttpPluginVersion = '^5.5.0';
