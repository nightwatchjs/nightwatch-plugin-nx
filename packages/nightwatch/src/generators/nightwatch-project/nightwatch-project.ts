import {
  addDependenciesToPackageJson,
  addProjectConfiguration,
  convertNxGenerator,
  formatFiles,
  generateFiles,
  getWorkspaceLayout,
  joinPathFragments,
  logger,
  names,
  offsetFromRoot,
  ProjectConfiguration,
  readProjectConfiguration,
  stripIndents,
  toJS,
  Tree,
  updateJson,
} from '@nrwl/devkit';
import { Linter } from '@nrwl/linter';
import { getRelativePathToRootTsConfig } from '@nrwl/workspace/src/utilities/typescript';

import { join } from 'path';
import { installedNightwatchVersion } from '../../utils/nightwatch-version';
import { filePathPrefix } from '../../utils/project-name';
import {
  nightwatchVersion,
  typesNodeVersion,
  typesNightwatchVersion,
  chromedriver,
  geckodriver,
} from '../../utils/versions';
// app
import { Schema } from './schema';

export interface NightwatchProjectSchema extends Schema {
  projectName: string;
  projectRoot: string;
}

function createFiles(tree: Tree, options: NightwatchProjectSchema) {
  const nightwatchVersion = installedNightwatchVersion();
  const nightwatchFiles = "v2"

  generateFiles(
    tree,
    join(__dirname, './files', nightwatchFiles),
    options.projectRoot,
    {
      tmpl: '',
      ...options,
      project: options.project || 'Project',
      ext: options.js ? 'js' : 'ts',
      offsetFromRoot: offsetFromRoot(options.projectRoot),
      rootTsConfigPath: getRelativePathToRootTsConfig(
        tree,
        options.projectRoot
      ),
    }
  );

  if (options.js) {
    toJS(tree);
  }
}

function addProject(tree: Tree, options: NightwatchProjectSchema) {
  let e2eProjectConfig: ProjectConfiguration;

  const detectedNighwatchVersion = installedNightwatchVersion() ?? nightwatchVersion;

  const nightwatchConfig = 'nightwatch.conf.js';

  if (options.baseUrl) {
    e2eProjectConfig = {
      root: options.projectRoot,
      sourceRoot: joinPathFragments(options.projectRoot, 'src'),
      projectType: 'application',
      targets: {
        e2e: {
          executor: 'nx:run-commands',
          options: {
            command: `npx nightwatch -c ${joinPathFragments(options.projectRoot, nightwatchConfig)} --baseURL=${options.baseUrl}`
          },
        },
      },
      tags: [],
      implicitDependencies: options.project ? [options.project] : undefined,
    };
  } else if (options.project) {
    const project = readProjectConfiguration(tree, options.project);

    if (!project.targets) {
      logger.warn(stripIndents`
      NOTE: Project, "${options.project}", does not have any targets defined and a baseUrl was not provided. Nx will use
      "${options.project}:serve" as the devServerTarget. But you may need to define this target within the project, "${options.project}".
      `);
    }
    const devServerTarget =
      project.targets?.serve && project.targets?.serve?.defaultConfiguration
        ? `${options.project}:serve:${project.targets.serve.defaultConfiguration}`
        : `${options.project}:serve`;
    e2eProjectConfig = {
      root: options.projectRoot,
      sourceRoot: joinPathFragments(options.projectRoot, 'src'),
      projectType: 'application',
      targets: {
        e2e: {
          executor: 'nx:run-commands',
          options: {
            command: `npx nightwatch -c ${joinPathFragments(options.projectRoot, nightwatchConfig)} --baseURL=${devServerTarget}`
          },
          configurations: {
            production: {
              devServerTarget: `${options.project}:serve:production`,
            },
          },
        },
      },
      tags: [],
      implicitDependencies: options.project ? [options.project] : undefined,
    };
  } else {
    throw new Error(`Either project or baseUrl should be specified.`);
  }

  if (detectedNighwatchVersion < 7) {
    e2eProjectConfig.targets.e2e.options.tsConfig = joinPathFragments(
      options.projectRoot,
      'tsconfig.json'
    );
  }
  addProjectConfiguration(
    tree,
    options.projectName,
    e2eProjectConfig,
    options.standaloneConfig
  );
}

export async function updateDependencies(tree: Tree) {
  return addDependenciesToPackageJson(
    tree,
    {},
    {
      nightwatch: nightwatchVersion,
      '@types/node': typesNodeVersion,
      '@types/nightwatch': typesNightwatchVersion,
      'geckodriver': geckodriver,
      'chromedriver': chromedriver,
    }
  );
}

export async function nightwatchProjectGenerator(host: Tree, schema: Schema) {
  const options = normalizeOptions(host, schema);
  const installTask = await updateDependencies(host);
  createFiles(host, options);
  addProject(host, options);
  if (!options.skipFormat) {
    await formatFiles(host);
  }
  return !options.skipPackageJson ? () => { installTask() } : () => { };
}

function normalizeOptions(host: Tree, options: Schema): NightwatchProjectSchema {
  const { appsDir } = getWorkspaceLayout(host);
  const projectName = filePathPrefix(
    options.directory ? `${options.directory}-${options.name}` : options.name
  );
  const projectRoot = options.directory
    ? joinPathFragments(
      appsDir,
      names(options.directory).fileName,
      options.name
    )
    : joinPathFragments(appsDir, options.name);

  options.linter = options.linter || Linter.EsLint;
  return {
    ...options,
    projectName,
    projectRoot,
  };
}

export default nightwatchProjectGenerator;
export const nightwatchProjectSchematic = convertNxGenerator(
  nightwatchProjectGenerator
);
