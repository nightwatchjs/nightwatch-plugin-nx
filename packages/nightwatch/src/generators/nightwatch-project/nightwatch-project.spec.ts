import {
  addProjectConfiguration,
  readJson,
  readProjectConfiguration,
  Tree,
  updateProjectConfiguration,
  WorkspaceJsonConfiguration,
} from '@nrwl/devkit';
import { createTreeWithEmptyV1Workspace } from '@nrwl/devkit/testing';
import { nightwatchProjectGenerator } from './nightwatch-project';
import { Schema } from './schema';
import { Linter } from '@nrwl/linter';
import { installedNightwatchVersion } from '../../utils/nightwatch-version';

jest.mock('../../utils/nightwatch-version');

describe('nightwatch Project', () => {
  let tree: Tree;
  const defaultOptions: Omit<Schema, 'name' | 'project'> = {
    linter: Linter.EsLint,
    standaloneConfig: false,
  };
  let mockedInstalledNightwatchVersion: jest.Mock<
    ReturnType<typeof installedNightwatchVersion>
  > = installedNightwatchVersion as never;

  beforeEach(() => {
    tree = createTreeWithEmptyV1Workspace();

    addProjectConfiguration(tree, 'my-app', {
      root: 'my-app',
      targets: {
        serve: {
          executor: 'serve-executor',
          options: {},
          configurations: {
            production: {},
          },
        },
      },
    });

    addProjectConfiguration(tree, 'my-dir-my-app', {
      root: 'my-dir/my-app',
      targets: {
        serve: {
          executor: 'serve-executor',
          options: {},
          configurations: {
            production: {},
          },
        },
      },
    });
  });
  afterEach(() => jest.clearAllMocks());

  describe('> v2', () => {
    beforeEach(() => {
      mockedInstalledNightwatchVersion.mockReturnValue(10);
    });

    it('should generate files for v10 and above', async () => {
      await nightwatchProjectGenerator(tree, {
        ...defaultOptions,
        name: 'my-app-e2e',
        project: 'my-app',
      });

      expect(tree.exists('apps/my-app-e2e/nightwatch.conf.ts')).toBeTruthy();

      expect(tree.exists('apps/my-app-e2e/src/test/login.ts')).toBeTruthy();
      expect(tree.exists('apps/my-app-e2e/src/test/titleAssertion.ts')).toBeTruthy();
      expect(tree.exists('apps/my-app-e2e/src/support/app.po.ts')).toBeTruthy();
    });

    it('should not add lint target when "none" is passed', async () => {
      await nightwatchProjectGenerator(tree, {
        name: 'my-app-e2e',
        project: 'my-app',
        linter: Linter.None,
        standaloneConfig: false,
      });
      const workspaceJson = readJson(tree, 'workspace.json');
      const project = workspaceJson.projects['my-app-e2e'];

      expect(project.architect.lint).toBeUndefined();
    });

    it('should update tags and implicit dependencies', async () => {
      await nightwatchProjectGenerator(tree, {
        name: 'my-app-e2e',
        project: 'my-app',
        linter: Linter.EsLint,
        standaloneConfig: false,
      });

      const project = readProjectConfiguration(tree, 'my-app-e2e');
      expect(project.tags).toEqual([]);
      expect(project.implicitDependencies).toEqual(['my-app']);
    });

    it('should set right path names in `nightwatch.config.ts`', async () => {
      await nightwatchProjectGenerator(tree, {
        ...defaultOptions,
        name: 'my-app-e2e',
        project: 'my-app',
      });
      const nightwatchConfig = tree.read(
        'apps/my-app-e2e/nightwatch.config.ts',
        'utf-8'
      );
      expect(nightwatchConfig).toMatchSnapshot();
    });

    it('should set right path names in `tsconfig.e2e.json`', async () => {
      await nightwatchProjectGenerator(tree, {
        ...defaultOptions,
        name: 'my-app-e2e',
        project: 'my-app',
      });
      const tsconfigJson = readJson(tree, 'apps/my-app-e2e/tsconfig.json');
      expect(tsconfigJson).toMatchSnapshot();
    });

    it('should extend from tsconfig.base.json', async () => {
      await nightwatchProjectGenerator(tree, {
        ...defaultOptions,
        name: 'my-app-e2e',
        project: 'my-app',
      });

      const tsConfig = readJson(tree, 'apps/my-app-e2e/tsconfig.json');
      expect(tsConfig.extends).toBe('../../tsconfig.base.json');
    });

    it('should support a root tsconfig.json instead of tsconfig.base.json', async () => {
      tree.rename('tsconfig.base.json', 'tsconfig.json');

      await nightwatchProjectGenerator(tree, {
        ...defaultOptions,
        name: 'my-app-e2e',
        project: 'my-app',
      });

      const tsConfig = readJson(tree, 'apps/my-app-e2e/tsconfig.json');
      expect(tsConfig.extends).toBe('../../tsconfig.json');
    });

    describe('nested', () => {
      it('should set right path names in `nightwatch.config.ts`', async () => {
        await nightwatchProjectGenerator(tree, {
          ...defaultOptions,
          name: 'my-app-e2e',
          project: 'my-dir-my-app',
          directory: 'my-dir',
        });

        const nightwatchConfig = tree.read(
          'apps/my-dir/my-app-e2e/nightwatch.conf.ts',
          'utf-8'
        );
        expect(nightwatchConfig).toMatchSnapshot();
      });

      it('should set right path names in `tsconfig.e2e.json`', async () => {
        await nightwatchProjectGenerator(tree, {
          ...defaultOptions,
          name: 'my-app-e2e',
          project: 'my-dir-my-app',
          directory: 'my-dir',
        });
        const tsconfigJson = readJson(
          tree,
          'apps/my-dir/my-app-e2e/tsconfig.json'
        );

        expect(tsconfigJson).toMatchSnapshot();
      });

      it('should extend from tsconfig.base.json', async () => {
        await nightwatchProjectGenerator(tree, {
          ...defaultOptions,
          name: 'my-app-e2e',
          project: 'my-app',
          directory: 'my-dir',
        });

        const tsConfig = readJson(tree, 'apps/my-dir/my-app-e2e/tsconfig.json');
        expect(tsConfig.extends).toBe('../../../tsconfig.base.json');
      });

      it('should support a root tsconfig.json instead of tsconfig.base.json', async () => {
        tree.rename('tsconfig.base.json', 'tsconfig.json');

        await nightwatchProjectGenerator(tree, {
          ...defaultOptions,
          name: 'my-app-e2e',
          project: 'my-app',
          directory: 'my-dir',
        });

        const tsConfig = readJson(tree, 'apps/my-dir/my-app-e2e/tsconfig.json');
        expect(tsConfig.extends).toBe('../../../tsconfig.json');
      });
    });

    describe('--project', () => {
      describe('none', () => {
        it('should not add any implicit dependencies', async () => {
          await nightwatchProjectGenerator(tree, {
            ...defaultOptions,
            name: 'my-app-e2e',
            baseUrl: 'http://localhost:7788',
          });

          const workspaceJson = readJson<WorkspaceJsonConfiguration>(
            tree,
            'workspace.json'
          );
          const projectConfig = workspaceJson.projects['my-app-e2e'];
          expect(projectConfig.implicitDependencies).not.toBeDefined();
          expect(projectConfig.tags).toEqual([]);
        });
      });

      it('should not throw an error when --project does not have targets', async () => {
        const projectConf = readProjectConfiguration(tree, 'my-app');
        delete projectConf.targets;

        updateProjectConfiguration(tree, 'my-app', projectConf);
        await nightwatchProjectGenerator(tree, {
          name: 'my-app-e2e',
          project: 'my-app',
          linter: Linter.EsLint,
        });

        const projectConfig = readProjectConfiguration(tree, 'my-app-e2e');
        expect(projectConfig.targets['e2e'].options.devServerTarget).toEqual(
          'my-app:serve'
        );
      });
    });

    it('should generate in the correct folder', async () => {
      await nightwatchProjectGenerator(tree, {
        ...defaultOptions,
        name: 'other-e2e',
        project: 'my-app',
        directory: 'one/two',
      });
      const workspace = readJson(tree, 'workspace.json');
      expect(workspace.projects['one-two-other-e2e']).toBeDefined();
      [
        'apps/one/two/other-e2e/nightwatch.conf.ts',
        'apps/one/two/other-e2e/src/test/login.ts',
      ].forEach((path) => expect(tree.exists(path)).toBeTruthy());
    });
  });
});
