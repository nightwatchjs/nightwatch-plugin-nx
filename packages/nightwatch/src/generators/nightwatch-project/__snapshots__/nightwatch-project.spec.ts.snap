// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nightwatch Project > v2 nested should set right path names in \`nightwatch.config.ts\` 1`] = `
"// Refer to the online docs for more details:
// https://nightwatchjs.org/gettingstarted/configuration/
//

//  _   _  _         _      _                     _          _
// | \\\\ | |(_)       | |    | |                   | |        | |
// |  \\\\| | _   __ _ | |__  | |_ __      __  __ _ | |_   ___ | |__
// | . \` || | / _\` || '_ \\\\ | __|\\\\ \\\\ /\\\\ / / / _\` || __| / __|| '_ \\\\
// | |\\\\  || || (_| || | | || |_  \\\\ V  V / | (_| || |_ | (__ | | | |
// \\\\_| \\\\_/|_| \\\\__, ||_| |_| \\\\__|  \\\\_/\\\\_/   \\\\__,_| \\\\__| \\\\___||_| |_|
//             __/ |
//            |___/

module.exports = {
  // An array of folders (excluding subfolders) where your tests are located;
  // if this is not specified, the test source must be passed as the second argument to the test runner.
  src_folders: ['apps/my-dir/my-app-e2e/src/test'],

  // See https://nightwatchjs.org/guide/concepts/page-object-model.html
  page_objects_path: ['apps/my-dir/my-app-e2e/src/support'],

  // See https://nightwatchjs.org/guide/extending-nightwatch/adding-custom-commands.html
  custom_commands_path: ['apps/my-dir/my-app-e2e/src/support'],

  // See https://nightwatchjs.org/guide/extending-nightwatch/adding-custom-assertions.html
  // custom_assertions_path: ['apps/my-dir/my-app-e2e/nightwatch/custom-assertions'],

  // See https://nightwatchjs.org/guide/extending-nightwatch/adding-plugins.html
  // plugins: ['@nightwatch/react'],

  // See https://nightwatchjs.org/guide/concepts/test-globals.html
  // globals_path: 'apps/my-dir/my-app-e2e/src/globals/globals.js',

  webdriver: {},

  test_workers: {
    enabled: true
  },
  test_settings: {
    default: {
      disable_error_log: false,
      launch_url: 'http://localhost:4200',

      screenshots: {
        enabled: false,
        path: 'screens',
        on_failure: true,
      },

      desiredCapabilities: {
        browserName: 'chrome',
      },

      webdriver: {
        start_process: true,
        server_path: '',
        cli_args: [
          // --verbose
        ],
      },
    },
    chrome: {
      desiredCapabilities: {
        browserName: 'chrome',
        'goog:chromeOptions': {
          // More info on Chromedriver: https://sites.google.com/a/chromium.org/chromedriver/
          //
          // w3c:false tells Chromedriver to run using the legacy JSONWire protocol (not required in Chrome 78)
          w3c: true,
          args: [
            //'--no-sandbox',
            //'--ignore-certificate-errors',
            //'--allow-insecure-localhost',
            //'--headless'
          ],
        },
      },
      firefox: {
        desiredCapabilities: {
          browserName: 'firefox',
          alwaysMatch: {
            acceptInsecureCerts: true,
            'moz:firefoxOptions': {
              args: [
                // '-headless',
                // '-verbose'
              ],
            },
          },
        },
        webdriver: {
          start_process: true,
          server_path: '',
          cli_args: [
            // very verbose geckodriver logs
            // '-vv'
          ],
        },
      },
      edge: {
        desiredCapabilities: {
          browserName: 'MicrosoftEdge',
          'ms:edgeOptions': {
            w3c: true,
            // More info on EdgeDriver: https://docs.microsoft.com/en-us/microsoft-edge/webdriver-chromium/capabilities-edge-options
            args: [
              //'--headless'
            ],
          },
        },
        webdriver: {
          start_process: true,
          // Follow https://docs.microsoft.com/en-us/microsoft-edge/webdriver-chromium/?tabs=c-sharp#download-microsoft-edge-webdriver
          // to download the Edge WebDriver and set the location of extracted \`msedgedriver\` below:
          server_path: '',
          cli_args: [
            // --verbose
          ],
        },
      },
    },
  },

  usage_analytics: {
    enabled: false,
    log_path: './logs/analytics',
    client_id: 'f9eed027-3a2d-4157-8cb5-d591acdc0095'
  }
};
"
`;

exports[`nightwatch Project > v2 nested should set right path names in \`tsconfig.e2e.json\` 1`] = `
Object {
  "compilerOptions": Object {
    "allowJs": true,
    "outDir": "../../../dist/out-tsc",
    "sourceMap": false,
    "types": Array [
      "nightwatch",
      "node",
    ],
  },
  "extends": "../../../tsconfig.base.json",
  "include": Array [
    "src/**/*.ts",
    "src/**/*.js",
    "nightwatch.conf.ts",
  ],
}
`;

exports[`nightwatch Project > v2 should set right path names in \`nightwatch.config.ts\` 1`] = `null`;

exports[`nightwatch Project > v2 should set right path names in \`tsconfig.e2e.json\` 1`] = `
Object {
  "compilerOptions": Object {
    "allowJs": true,
    "outDir": "../../dist/out-tsc",
    "sourceMap": false,
    "types": Array [
      "nightwatch",
      "node",
    ],
  },
  "extends": "../../tsconfig.base.json",
  "include": Array [
    "src/**/*.ts",
    "src/**/*.js",
    "nightwatch.conf.ts",
  ],
}
`;
